src/passes/Outlining.cpp

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ir/module-utils.h"
+#include "ir/names.h"
+#include "ir/utils.h"
+#include "pass.h"
+#include "stringify-walker.h"
+#include "wasm-builder.h"
+#include "wasm-stack.h"
+#include "wasm.h"
+
+namespace wasm {
+
+/*
+ * This walker performs a shallow visit of control-flow (try, if, block, loop)
+ * expressions and their simple expression siblings before then visiting the
+ * children of each control-flow expression in postorder. As a result, this
+ * walker un-nests nested control flow structures, so the expression visit order
+ * does not correspond to a normal postorder traversal of the function.
+ *
+ * For example, the below (contrived) wat:
+ * 1: (block
+ * 2:   (i32.add
+ * 3:     (i32.const 20)
+ * 4:     (i32.const 10))
+ * 5:   (if
+ * 6:     (i32.const 0)
+ * 7:     (then (return (i32.const 1)))
+ * 8:     (else (return (i32.const 0)))
+ * 9:   )
+ *
+ * Would have it's expressions visited in the following order (based on line
+ * number):
+ * 1, 3, 4, 2, 6, 5, 7, 8
+ *
+ * Of note:
+ *   - The add (line 2) binary operator's left and right children are visited
+ * first as they need to be on the stack before the add operation is executed
+ *   - The if-condition (i32.const 0) on line 6 is visited before the if
+ * expression
+ */
+template<typename SubType>
+void StringifyWalker<SubType>::walkModule(SubType* self, Module* module) {
+  self->wasm = module;
+  ModuleUtils::iterDefinedFunctions(*module, [&](Function* func) {
+    /*
+     * The ordering of the below lines of code are important. On each function
+     * iteration, we:
+     * 1. push a task for calling the handler function, to ensure that each
+     *    function has an opportunity to dequeue from StringifyWalker's
+     *    internally managed queue. This queue exists to provide a way for
+     *    control flow to defer scanning their children.
+     * 2. push a task for adding a unique symbol, so that after the function
+     *    body is visited as a single expression, there is a a separator between
+     *    the symbol for the entire function and the symbol for the function's
+     *    children
```

Comment:
I mean children of the function block here. We evaluate the function block as a unit or single expression, add a unique symbol, and then visit the function's children. Perhaps I can reword.

Reply:
I think this is confusing because I interpret "function's children" as the children of the root expression, but if the `root` is e.g. an `i32.add`, then its children are visited _before_ the unique symbol this comment is referring to is emitted.

Maybe it would be clearer to  "control flow contents" instead of "control flow children." Then "control flow contents" would mean expressions inside a control flow structure and "children" would refer to all other expressions.

Reply:
done